go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,317 @@
+			}
+		case mcap.TokenMessage:
+			message, err := mcap.ParseMessage(data)
+			if err != nil {
+				return err
+			}
+			if message.LogTime < opts.start {
+				continue
+			}
+			if message.LogTime >= opts.end {
+				continue
+			}
+			channel, ok := channels[message.ChannelID]
+			if !ok {
+				continue
+			}
+			if !channel.written {
+				schema, ok := schemas[channel.SchemaID]
+				if !ok {
+					return fmt.Errorf("encountered channel with topic %s with unknown schema ID %d", channel.Topic, channel.SchemaID)
+				}
+				if !schema.written {
+					if err = mcapWriter.WriteSchema(schema.Schema); err != nil {
+						return err
+					}
+					schema.written = true
+				}
+				if err = mcapWriter.WriteChannel(channel.Channel); err != nil {
+					return err
+				}
+				channel.written = true
+			}
+			if err = mcapWriter.WriteMessage(message); err != nil {
+				return err
+			}
+		case mcap.TokenAttachment:
+			if !opts.includeAttachments {
+				continue
+			}
+			attachment, err := mcap.ParseAttachment(data)
+			if err != nil {
+				return err
+			}
+			if attachment.LogTime < opts.start {
+				continue
+			}
+			if attachment.LogTime >= opts.end {
+				continue
+			}
+			if err = mcapWriter.WriteAttachment(attachment); err != nil {
+				return err
+			}
+		case mcap.TokenMetadata:
+			if !opts.includeMetadata {
+				continue
+			}
+			metadata, err := mcap.ParseMetadata(data)
+			if err != nil {
+				return err
+			}
+			if err = mcapWriter.WriteMetadata(metadata); err != nil {
+				return err
+			}
+		case mcap.TokenDataEnd, mcap.TokenFooter:
+			// data section is over, either because the file is over or the summary section starts.
+			return nil
+		case mcap.TokenChunk:
+			return errors.New("expected lexer to remove chunk records from input stream")
+		case mcap.TokenError:
+			return errors.New("received error token but lexer did not return error on Next")
+		}
+	}
+}
+
+func init() {
+	rootCmd.AddCommand(filterCmd)
+
+	filterCmd.PersistentFlags().StringVarP(&filterOutput, "output", "o", "", "output filename")
+	filterCmd.PersistentFlags().StringArrayVarP(&filterIncludeTopics, "include-topic-regex", "y", []string{}, "messages with topic names matching this regex will be included")
+	filterCmd.PersistentFlags().StringArrayVarP(&filterExcludeTopics, "exclude-topic-regex", "n", []string{}, "messages with topic names matching this regex will be included")
+	filterCmd.PersistentFlags().Uint64VarP(&filterStart, "start-time-ns", "s", 0, "messages with log times after or equal to this timestamp will be included.")
+	filterCmd.PersistentFlags().Uint64VarP(&filterEnd, "end-time-ns", "e", math.MaxUint64, "messages with log times before timestamp will be included.")
+	filterCmd.PersistentFlags().BoolVar(&filterIncludeMetadata, "include-metadata", true, "whether to include metadata in the output bag")
+	filterCmd.PersistentFlags().BoolVar(&filterIncludeAttachments, "include-attachments", true, "whether to include attachments in the output mcap")
```

Comment:
I think we should probably default both include-metadata and include-attachments to false - since in the common case of filtering for a single topic, there's no logical reason why all your metadata/attachments should relate to that topic